kernel/src/log_segment.rs

```diff
@@ -57,126 +70,93 @@ pub(crate) struct LogSegment {
+    pub(crate) fn build_complete(self, storage: &dyn StorageHandler) -> DeltaResult<LogSegment> {
+        let checkpoint_hint = LastCheckpointHint::try_read(storage, &self.log_root)?;
+        self.build_complete_impl(storage, checkpoint_hint)
+    }
 
-        let listed_files = match (checkpoint_hint.into(), time_travel_version) {
+    // factored out for testing
+    fn build_complete_impl(
+        self,
+        storage: &dyn StorageHandler,
+        last_checkpoint_hint: Option<LastCheckpointHint>,
+    ) -> DeltaResult<LogSegment> {
+        let listed_files = match (last_checkpoint_hint, self.end_version) {
             (Some(cp), None) => {
-                ListedLogFiles::list_with_checkpoint_hint(&cp, storage, &log_root, None)?
+                ListedLogFiles::list_with_checkpoint_hint(&cp, storage, &self.log_root, None)?
             }
             (Some(cp), Some(end_version)) if cp.version <= end_version => {
                 ListedLogFiles::list_with_checkpoint_hint(
                     &cp,
                     storage,
-                    &log_root,
+                    &self.log_root,
                     Some(end_version),
                 )?
             }
-            _ => ListedLogFiles::list(storage, &log_root, None, time_travel_version)?,
+            _ => ListedLogFiles::list(storage, &self.log_root, None, self.end_version)?,
         };
 
-        LogSegment::try_new(listed_files, log_root, time_travel_version)
+        LogSegment::try_new(listed_files, self.log_root, self.end_version)
     }
 
-    /// Constructs a [`LogSegment`] to be used for `TableChanges`. For a TableChanges between versions
-    /// `start_version` and `end_version`: Its LogSegment is made of zero checkpoints and all commits
-    /// between versions `start_version` (inclusive) and `end_version` (inclusive). If no `end_version`
-    /// is specified it will be the most recent version by default.
+    /// Constructs a incremental [`LogSegment`]. It is 'incremental' in the sense that it does
+    /// _not_ include all actions on a table, but only the actions between two versions [start,
+    /// end].
+    ///
+    /// This `LogSegment` is made of zero checkpoints and all commits between the specified start
+    /// and end versions (both inclusive).
+    ///
+    /// If no `end_version` is specified, we will LIST to determine the most recent version.
+    /// Start version is required and is inclusive.
     #[internal_api]
-    pub(crate) fn for_table_changes(
-        storage: &dyn StorageHandler,
-        log_root: Url,
+    pub(crate) fn build_incremental(
+        self,
         start_version: Version,
-        end_version: impl Into<Option<Version>>,
-    ) -> DeltaResult<Self> {
-        let end_version = end_version.into();
-        if let Some(end_version) = end_version {
+        storage: &dyn StorageHandler,
+    ) -> DeltaResult<LogSegment> {
+        if let Some(end_version) = self.end_version {
```

Comment:
One way to unify timestamp conversion and CDF is to instead specify a lower bound:
```rust
enum IncrementalLogSegmentBound {
    // Specifies the max number of commits from the end.
    Limit(usize), 
    // Specifies the starting commit of the range
    StartingFrom(Version)
}
```

This would look like:
```rust
builder.build_incremental(&storageHandler, StartingFrom(5))
// or 
builder.build_incremental(&storageHandler, Limit(10))
```